hack/upgrade-test-harness/main.go

```diff
@@ -0,0 +1,186 @@
+	opts                   = configOpts{}
+)
+
+func main() {
+	cmd := &cobra.Command{
+		Use:           "eck-upgrade-harness",
+		Short:         "Test harness for testing ECK release upgrades",
+		Version:       "0.1.0",
+		SilenceErrors: true,
+		SilenceUsage:  true,
+		RunE:          doRun,
+	}
+
+	cmd.Flags().StringVar(&opts.confFile, "conf-file", "conf.yaml", "Path to the file containing test params")
+	cmd.Flags().StringVar(&opts.fromRelease, "from-release", "alpha", "Release to start with (alpha, beta, ga101, ga112, upcoming)")
+	cmd.Flags().StringVar(&opts.logLevel, "log-level", "INFO", "Log level (DEBUG, INFO, WARN, ERROR)")
+	cmd.Flags().UintVar(&opts.retryCount, "retry-count", 5, "Number of retries")
+	cmd.Flags().DurationVar(&opts.retryDelay, "retry-delay", 30*time.Second, "Delay between retries")
+	cmd.Flags().DurationVar(&opts.retryTimeout, "retry-timeout", 300*time.Second, "Time limit for retries")
+	cmd.Flags().BoolVar(&opts.skipCleanup, "skip-cleanup", false, "Skip cleaning up after test run")
+	cmd.Flags().StringVar(&opts.toRelease, "to-release", "upcoming", "Release to finish with (alpha, beta, ga101, ga112, upcoming)")
+	cmd.Flags().StringVar(&opts.upcomingReleaseYAML, "upcoming-release-yaml", "../../config/all-in-one.yaml", "YAML file for installing the upcoming release")
+
+	kubeConfFlags.AddFlags(cmd.Flags())
+
+	if err := cmd.Execute(); err != nil {
+		zap.S().Errorw("Test failed", "error", err)
+		os.Exit(1)
+	}
+}
+
+func doRun(_ *cobra.Command, _ []string) error {
+	config.InitLogging(opts.logLevel)
+
+	// load test configurations
+	conf, err := config.Load(opts.confFile)
+	if err != nil {
+		return fmt.Errorf("failed to load config: %w", err)
+	}
+
+	// determine the release to test
+	from := mustGetReleasePos(conf, opts.fromRelease)
+	to := mustGetReleasePos(conf, opts.toRelease)
+
+	if to < from {
+		return errInvalidVersionRange
+	}
+
+	// create test context
+	ctx, err := fixture.NewTestContext(kubeConfFlags, int(opts.retryCount), opts.retryDelay, opts.retryTimeout)
+	if err != nil {
+		return err
+	}
+
+	// configure cleanup
+	if !opts.skipCleanup {
+		defer func() {
+			ctx.Info("Starting cleanup")
+
+			if err := ctx.Cleanup(ctx); err != nil {
+				ctx.Errorw("Failed to cleanup", "error", err)
+			}
+		}()
+	}
+
+	// run upgrade fixtures
+	var prevTestParam *fixture.TestParam
+
+	for i := from; i <= to; i++ {
+		currTestParam := conf.TestParam[i]
+
+		// setup upcoming release if necessary
+		if currTestParam.Name == "upcoming" {
+			if err := setupUpcomingRelease(opts.upcomingReleaseYAML); err != nil {
+				return fmt.Errorf("failed to setup upcoming release: %w", err)
+			}
+		}
+
+		fixtures := buildUpgradeFixtures(prevTestParam, currTestParam)
+
+		ctx.Infof("=====[%s]=====", currTestParam.Name)
+
+		for _, f := range fixtures {
+			if err := f.Execute(ctx); err != nil {
+				return err
+			}
+		}
+
+		prevTestParam = &currTestParam
+	}
+
+	// try to scale the last deployed release
+	if prevTestParam != nil {
```

Comment:
It's not clear to me why scaling is only happening at the end.

Reply:
I think the purpose of the test is to make sure that the operator works correctly after going through several in-place upgrades so the scaling is done only at the end. Scaling on every upgrade step doesn't add much value because they are past releases and it's a bit too late to fix any problems they might have anyway.